Use mktemp

rockandska opened the pull request:
- [X] I have reviewed the [Contributor Guidelines][contributor].
- [X] I have reviewed the [Code of Conduct][coc] and agree to abide by it

[contributor]: https://github.com/bats-core/bats-core/blob/master/docs/CONTRIBUTING.md
[coc]:         https://github.com/bats-core/bats-core/blob/master/docs/CODE_OF_CONDUCT.md

Using mktemp is safer than using $PID and avoid race condition, especially when running inside docker because the PID is rarely != 1.
That PR allows to running multiples times the same tests safely with `--no-tempdir-cleanup` and avoid race condition too.
Without this PR we hit on the second run and is quite annoying :

```
Error: BATS_RUN_TMPDIR (/tmp/bats-run-1) already exists
Reusing old run directories can lead to unexpected results ... aborting!
```

As an example for debugging :
```bash
$ cat test.sh
mkdir -p "$PWD/tmp"
for BASHVER in 3.2 4.1 4.4;do
	docker build --build-arg bashver="${BASHVER}" --tag "bats/bats:bash-${BASHVER}" . || exit $?
	docker run -it "bash:${BASHVER}" --version || exit $?
	time docker run \
          -u "$(id -u):0" \
          -it \
          -v "/etc/passwd:/etc/passwd:ro" \
          -v "${PWD}/tmp:/tmp" \
          "bats/bats:bash-${BASHVER}" \
          --tap /opt/bats/test \
          --no-tempdir-cleanup \
          ${@:+--filter "${@}"} \
          || exit $?
done
$ ./test.sh '.*tmpdir.*'
```
Result after some run of bats core self tests :
```
$ ls -1 tmp/
bats-run-1-Jamgap
bats-run-1-KGLIOm
bats-run-1-kJLIJD
bats-run-1-NdbhKe
bats-run-1-npAgfd
bats-run-1-oklMGC
bats-run-test-tmpdir-9354-DPMCIb
bats-run-test-tmpdir-9354-kgppIg
bats-run-test-tmpdir-9354-ooIgGd
bats-run-test-tmpdir-9355-EkNegL
bats-run-test-tmpdir-9410-fbLkJa
bats-run-test-tmpdir-9410-gBFNEK
bats-run-test-tmpdir-9410-PBifMi
bats-run-test-tmpdir-9411-kclEOi
bats-run-test-tmpdir-9481-dPLPNB
bats-run-test-tmpdir-9481-elDdeH
bats-run-test-tmpdir-9481-nLAEbE
bats-run-test-tmpdir-9482-DEFDdi
bats-run-test-tmpdir-9540-cbCCdl
bats-run-test-tmpdir-9542-ngkhMF
bats-run-test-tmpdir-9597-mFaEac
bats-run-test-tmpdir-9599-lEaffJ
bats-run-test-tmpdir-9669-cpmdHK
bats-run-test-tmpdir-9671-CaipNk
```

martin-schulze-vireso commented on libexec/bats-core/bats:
```diff
@@ -200,12 +200,19 @@ while [[ "$#" -ne 0 ]]; do
-  printf "Error: BATS_RUN_TMPDIR (%s) already exists\n" "$BATS_RUN_TMPDIR" >&2
-  printf "Reusing old run directories can lead to unexpected results ... aborting!\n" >&2
-  exit 1
+if [[ -n "${BATS_RUN_TMPDIR:-}" ]];then
```
So far, we don't allow for setting `BATS_RUN_TMPDIR` from outside. It is already customizable via `BATS_TMPDIR`, so I think we don't need this additional complexity here.

btamayo replied:
Do you think having this check (for `-n "${BATS_RUN_TMPDIR:-}") here anyway could potentially expose and add a guardrail against race conditions?

martin-schulze-vireso replied:
It would allow an opportunity for better reporting in cases where it is detectable but I don't think it can eliminate race conditions entirely, since the check and `mktemp` are not one atomic operation.

btamayo commented on docs/source/writing-tests.md:
```diff
@@ -305,6 +305,9 @@ There are several global variables you can use to introspect on Bats tests:
 - `$BATS_TEST_NUMBER` is the (1-based) index of the current test case in the test file.
 - `$BATS_SUITE_TEST_NUMBER` is the (1-based) index of the current test case in the test suite (over all files).
 - `$BATS_TMPDIR` is the location to a directory that may be used to store temporary files.
+- `$BATS_RUN_TMPDIR` is the location to the temporary directory used by bats to
```
I propose ([here](https://github.com/bats-core/bats-core/pull/410/files#r575897751) as well) we move internal variables to a separate section to minimize folks using them without breaking existing logic.

btamayo commented:
I think we can add this to 1.3.1. @rockandska Thank you for this. Can you resolve the conflicts and address the comment(s). Otherwise, LGTM.

rockandska commented:
@martin-schulze-vireso / @btamayo :

- add an error and exit if mktemp failed
- moved BATS_RUN_TMPDIR to an `internal variables section`
- rebased with master
